File: LinkedList.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = next

class Linked_List:
    def __init__(self):
        self.head = None

    def insertAtBeginning(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def insertAtTheEnd(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        current_node = self.head
        while current_node.next:
            current_node = current_node.next
        current_node.next = new_node

    def insertAtTheIndex(self, data, index):
        if index == 0:
            insertAtBeginning(data)
            return

        index_at_the_moment = 0
        current_node = self.head
        while index_at_the_moment > index and current_node is not None:
            index_at_the_moment += 1
            current_node = current_node.next

        if current_node is not None:
            new_node = Node(data)
            new_node.next = current_node.next
            current_node.next = new_node
        else:
            print("Nie ma takiego Indexu")

    def updateNode(self, val, index):
        current_node = self.head
        index_at_the_moment = 0
        while index_at_the_moment != index:
            current_node = current_node.next
            index_at_the_moment += 1
        if current_node is not None:
            current_node.value = val
        else:
            print("Nie ma takiego Indexu")

    def removeFirstNode(self):
        if self.head is None:
            print("Nie ma pierwszego elemntu Listy Linkowanej")
            return
        self.head = self.head.next

    def removeLastNode(self):
        if self.head is None:
            print("Nie ma pierwszego elemntu Listy Linkowanej")
            return

        current_node = self.head
        while current_node.next and current_node.next.next:
            current_node = current_node.next

        current_node.next = None

    def size(self):
        size = 0
        current_node = self.head
        while current_node:
            size += 1
            current_node = current_node.next
        return size

    def print(self):
        nice_list = ""
        current_node = self.head
        while current_node is not None:
            nice_list += current_node.data + "-->"
            current_node = current_node.next
        return nice_list



linked_list = Linked_List()
linked_list.insertAtBeginning("5")
linked_list.insertAtTheEnd("2")
linked_list.size()
linked_list.print()

File: LinkedList.js
class Node {
  constructor(data) {
    this.next = null;
    this.data = data;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
  }

  addAtTheStart(data) {
    let node = new Node(data);
    if (this.head === null) {
      this.head = node;
    } else {
      node.next = this.head;
      this.head = node;
    }
  }

  addAtTheEnd(data) {
    let node = new Node(data);
    if (this.head === null) {
      this.head = node;
    }
    let currentNode = this.head;
    while (currentNode.next) {
      currentNode = currentNode.next;
    }
    currentNode.next = node;
  }

  addAtTheIndex(data, index) {
    let node = new Node(data);
    let indexNow = 0;
    let currentNode = this.head;
    if (index < 0 || index > (this.lenght() + 1)) {
      console.log("Nie poprawny index")
      return;
    }
    while (currentNode.next && index > indexNow) {
      currentNode = currentNode.next;
    }
    currentNode.next = node;
  }

  deleteAtTheStart() {
    if (this.head === null) {
      return
    }
    this.head = this.head.next;
  }

  deleteAtTheEnd() {
    let currentNode = this.head;
    if (this.head === null) {
      return;
    }
    if (this.head.next === null) {
      this.head = null;
      return;
    }
    while (currentNode.next.next) {
      currentNode = currentNode.next;
    }
    currentNode.next = null;
  }

  deleteAtTheIndex(index) {
    if (this.head === null) {
      return "Lista jest pusta";
    }
    if (index < 0) {
      return "Index nie moze byc mniejszy niz 0";
    }
    if (index === 0) {
      this.deleteAtTheStart();
    }
    let currentNode = this.head;
    let indexAtTheMoment = 0;
    let previousNode = null;
    while (currentNode && indexAtTheMoment < index) {
      previousNode = currentNode;
      currentNode = currentNode.next;
      indexAtTheMoment += 1;
    }
    if (currentNode === null) {
      return "nie ma takiego indexu"
    }

    if (previousNode !== null) {
      previousNode.next = currentNode.next;
    }
  }

  findIndexByValue(value) {
    let index = 0;
    let currentNode = this.head;
    while (currentNode) {
      if (currentNode.data == value) {
        return index;
      }
      index += 1;
      currentNode = currentNode.next;
    }
    return "Nie znaleziono takiej wartosci";
  }

  findValueByIndex(index) {
    let indexAtTheMoment = 0;
    let currentNode = this.head;
    while (currentNode) {
      if (indexAtTheMoment == index) {
        return currentNode.data
      }
      indexAtTheMoment += 1
      currentNode = currentNode.next
    }
    return "Nie znaleziono takiego indexu"
  }


  lenght() {
    let len = 0;
    let currentNode = this.head;
    while (currentNode) {
      len += 1
      currentNode = currentNode.next;
    }
    return len;
  }

  logIt() {
    let str = "";
    let currentNode = this.head;
    while (currentNode) {
      if (currentNode.next === null) {
        str += currentNode.data;
        break;
      }
      str += currentNode.data + "-->";
      currentNode = currentNode.next;
    }
    console.log(str);
  }
}

let LL = new LinkedList

LL.addAtTheStart(8)
LL.addAtTheStart(12)
LL.addAtTheEnd(16)
LL.addAtTheIndex(21, 2)
LL.logIt()
LL.deleteAtTheStart()
LL.deleteAtTheEnd()
LL.logIt()
console.log(LL.findIndexByValue(16))
console.log(LL.findIndexByValue(213))
console.log(LL.findValueByIndex(0))
console.log(LL.findValueByIndex(20))
console.log(LL.lenght())
LL.addAtTheStart(3)
LL.addAtTheEnd(333)
LL.logIt()
console.log(LL.deleteAtTheIndex(-2))
LL.deleteAtTheIndex(0)
LL.deleteAtTheIndex(2)
console.log(LL.deleteAtTheIndex(2000))
LL.logIt()


